README.md
<p align="center"><img src="https://github.com/hacktons/convex_bottom_bar/raw/master/doc/preview.png" alt="appBar preview"></p>
<p align="center">
  <a href="https://pub.dev/packages/convex_bottom_bar_gothomesafe"><img src="https://img.shields.io/pub/v/convex_bottom_bar_gothomesafe.svg" alt="pub.dev"></a>
  <a href="https://pub.dev/packages/convex_bottom_bar_gothomesafe/versions/3.0.0-nullsafety.0"><img src="https://img.shields.io/badge/nullsafety-3.0.0-blue.svg" alt="pub.dev"></a>
  <a href="https://github.com/hacktons/convex_bottom_bar"><img src="https://img.shields.io/badge/platform-flutter-ff69b4.svg" alt="github"></a>
  <a href="https://coveralls.io/github/hacktons/convex_bottom_bar"><img src="https://coveralls.io/repos/github/hacktons/convex_bottom_bar/badge.svg" alt="coverage status"></a>
  <a href="https://pub.dev/packages/convex_bottom_bar_gothomesafe/score"><img src="https://badges.bar/convex_bottom_bar_gothomesafe/likes" alt="likes"></a>
  <a href="https://codemagic.io/apps/5db10f597d3edb001a6ede16/5db10f597d3edb001a6ede15/latest_build"><img src="https://api.codemagic.io/apps/5db10f597d3edb001a6ede16/5db10f597d3edb001a6ede15/status_badge.svg" alt="build status"></a>
  <a href="https://github.com/hacktons/convex_bottom_bar/raw/LICENSE"><img src="https://img.shields.io/github/license/hacktons/convex_bottom_bar.svg" alt="license"></a>
</p>
<p align="center">
    <a href="https://github.com/hacktons/convex_bottom_bar/blob/master/README.md">English</a>
    | <a href="https://github.com/hacktons/convex_bottom_bar/blob/master/README-zh.md">简体中文</a>
</p>

---

The official BottomAppBar can only display a notch FAB with an app bar, and sometimes we need a convex FAB. BottomAppBar and NotchShape's implementation inspires this ConvexAppBar.

Online example can be found at [https://appbar.codemagic.app](https://appbar.codemagic.app).

**convex_bottom_bar_gothomesafe is now a [Flutter Favorite](https://flutter.dev/docs/development/packages-and-plugins/favorites) package!**

<a href="https://flutter.dev/docs/development/packages-and-plugins/favorites">
<img height="128" src="https://github.com/hacktons/convex_bottom_bar/raw/master/doc/flutter-favorite.png">
</a>

Here are some supported style:

|             **fixed**             |            **react**             |      **badge chip**       |
| :-------------------------------: | :------------------------------: | :-----------------------: |
|     ![](doc/appbar-fixed.gif)     |    ![](doc/appbar-react.gif)     | ![](doc/appbar-badge.gif) |
|          **fixedCircle**          |         **reactCircle**          |         **flip**          |
| ![](doc/appbar-fixed-circle.gif)  | ![](doc/appbar-react-circle.gif) | ![](doc/appbar-flip.gif)  |
|            **textIn**             |            **titled**            |       **tab image**       |
|    ![](doc/appbar-textIn.gif)     |    ![](doc/appbar-titled.gif)    | ![](doc/appbar-image.gif) |
|            **button**             |         **fixed corner**         |                           |
| ![](doc/appbar-single-button.png) | ![](doc/appbar-corner-fixed.png) |                           |

## How to use

Typically ConvexAppBar can work with `Scaffold` by setup its `bottomNavigationBar`.

The `ConvexAppBar` has two constructors. The `ConvexAppBar()` will use the default style to simplify the tab creation.

Add this to your package's pubspec.yaml file, use the latest version [![Pub](https://img.shields.io/pub/v/convex_bottom_bar_gothomesafe.svg)](https://pub.dartlang.org/packages/convex_bottom_bar_gothomesafe):

```yaml
dependencies:
  convex_bottom_bar_gothomesafe: ^latest_version
```

```dart
import 'package:convex_bottom_bar_gothomesafe/convex_bottom_bar_gothomesafe.dart';

Scaffold(
  bottomNavigationBar: ConvexAppBar(
    items: [
      TabItem(icon: Icons.home, title: 'Home'),
      TabItem(icon: Icons.map, title: 'Discovery'),
      TabItem(icon: Icons.add, title: 'Add'),
      TabItem(icon: Icons.message, title: 'Message'),
      TabItem(icon: Icons.people, title: 'Profile'),
    ],
    initialActiveIndex: 2,//optional, default as 0
    onTap: (int i) => print('click index=$i'),
  )
);
```

**Flutter Version Support**  
As Flutter is developing fast. There can be breaking changes. We will be trying to support the
stable version and beta version through different package versions.

| **Stable Flutter Version** | **Package Version** |                                  **More**                                  |
| :------------------------: | :-----------------: | :------------------------------------------------------------------------: |
|           >=1.20           |       >=2.4.0       |           Since v1.20, the stable version changed the Stack api            |
|           <1.20            |       <=2.3.0       | Support for stable version such as v1.17, v1.12 is not going to be updated |

## Features

- Provide multiple internal styles
- Ability to change the theme of AppBar
- Provide builder API to customize a new style
- Add badge on the tab menu
- Elegant transition animation
- Provide hook API to override some of the internal styles
- RTL support

## Table of contents

- [Theming](#theming)
- [Badge](#badge)
- [Single Button](#single-button)
- [Style Hook](#style-hook)
- [RTL Support](#rtl-support)
- [Custom Example](#custom-example)
- [FAQ](#faq)
- [Donate](#donate)

## Theming

The bar will use default style, you may want to theme it. Here are some supported attributes:

![](doc/appbar-theming.png)

| Attributes      | Description                                                                          |
| --------------- | ------------------------------------------------------------------------------------ |
| backgroundColor | AppBar background                                                                    |
| gradient        | gradient will override backgroundColor                                               |
| height          | AppBar height                                                                        |
| color           | tab icon/text color                                                                  |
| activeColor     | tab icon/text color **when selected**                                                |
| curveSize       | size of the convex shape                                                             |
| top             | top edge of the convex shape relative to AppBar                                      |
| cornerRadius    | draw the background with topLeft and topRight corner; Only work with fixed tab style |
| style           | style to describe the convex shape: **fixed, fixedCircle, react, reactCircle**, ...  |
| chipBuilder     | custom badge builder, use **ConvexAppBar.badge** for default badge                   |

## Badge

If you need to add a badge on the tab, use the `ConvexAppBar.badge` to get it done.

[![badge demo](doc/badge-demo-preview.gif)](doc/badge-demo.mp4 "badge demo")

```dart
ConvexAppBar.badge({0: '99+', 1: Icons.assistant_photo, 2: Colors.redAccent},
  items: [
    TabItem(icon: Icons.home, title: 'Home'),
    TabItem(icon: Icons.map, title: 'Discovery'),
    TabItem(icon: Icons.add, title: 'Add'),
  ],
  onTap: (int i) => print('click index=$i'),
);
```

The `badge()` method accepts an array of badges; The `badges` is a map with tab items. Each value of entry can be either `String`, `IconData`, `Color` or `Widget`.

## Single Button

If you only need a single button, checkout the `ConvexButton`.

![button](doc/appbar-single-shape.png)

```dart
Scaffold(
  appBar: AppBar(title: const Text('ConvexButton Example')),
  body: Center(child: Text('count $count')),
  bottomNavigationBar: ConvexButton.fab(
    onTap: () => setState(() => count++),
  ),
);
```

## Style Hook

Hook for internal tab style. Unlike the `ConvexAppBar.builder`, you may want to update the tab style without defining a new tab style.

**Warning:**  
This hook is limited and can lead to `overflow broken` if the size you provide does not match with internal style.

```dart
StyleProvider(
  style: Style(),
  child: ConvexAppBar(
    initialActiveIndex: 1,
    height: 50,
    top: -30,
    curveSize: 100,
    style: TabStyle.fixedCircle,
    items: [
      TabItem(icon: Icons.link),
      TabItem(icon: Icons.import_contacts),
      TabItem(title: "2020", icon: Icons.work),
    ],
    backgroundColor: _tabBackgroundColor,
  ),
)
class Style extends StyleHook {
  @override
  double get activeIconSize => 40;

  @override
  double get activeIconMargin => 10;

  @override
  double get iconSize => 20;

  @override
  TextStyle textStyle(Color color) {
    return TextStyle(fontSize: 20, color: color);
  }
}
```

## RTL Support

RTL is supported internally, and if you define the TextDirection inside the app, the AppBar should work fine.
Both RTL and LTR can be configured through `Directionality`:

```dart
Directionality(
  textDirection: TextDirection.rtl,
  child: Scaffold(body:ConvexAppBar(/*TODO ...*/)),
)
```

## Custom Example

If the default style does not match your situation, try with `ConvexAppBar.builder()`, allowing you to custom nearly all the tab features.

```dart
Scaffold(
  bottomNavigationBar: ConvexAppBar.builder(
    count: 5,
    backgroundColor: Colors.blue,
    style: TabStyle.fixed,
    itemBuilder: Builder(),
  )
);

// user defined class
class Builder extends DelegateBuilder {
  @override
  Widget build(BuildContext context, int index, bool active) {
    return Text('TAB $index');
  }
}
```

Full custom example can be found at [example](example).

## FAQ

Please file feature requests and bugs at the [issue tracker](https://github.com/hacktons/convex_bottom_bar/issues).

- [How to block tab event?](doc/how-to-block-tab-event.md)
- [Crash on flutter dev/beta channel](doc/issue-crash-on-flutter-dev-channel.md)
- [Change active tab index programmatically](doc/issue-change-active-tab-index.md)
- [Using an image instead of an icon for actionItem](doc/issue-image-for-actionitem.md)
- [Is there anyway to remove elevation in the bottom bar?](doc/issue-remove-elevation.md)

## Donate

You like the package ? Buy me a coffee :)

[![ko-fi](doc/donate-kofi1.png)](https://ko-fi.com/hacktons)


CHANGELOG.md
## [3.0.0-nullsafety.1]

- Prepare for Flutter 2.

## [2.7.1+2]

- format

## [2.7.1+1]

- Bug fix [Manual update tab index workaround seems not working since version 2.7.x](https://github.com/hacktons/convex_bottom_bar/issues/134)

## [2.7.1]

- fix runtime exception when there is no controller passed in

## [2.7.0+1]

- add new configuration `disableDefaultTabController`

## [3.0.0-nullsafety.0]

- Migrate to [null-safety](https://dart.dev/null-safety/migration-guide).

## [2.6.0]

- Fix tab highlight state to satisfy keyboard and router case; [#115](https://github.com/hacktons/convex_bottom_bar/issues/115), [#112](https://github.com/hacktons/convex_bottom_bar/issues/112)
- Breaking changes: disable initial index property when working with TabController to avoid potential index conflict;

## [2.5.1+1]

- Bug fix [#111 Tab selection not registered during tab transition](https://github.com/hacktons/convex_bottom_bar/issues/111)

## [2.5.1]

- Replace all Stack.overflow to Stack.clipBehavior.
- The removal of [Stack.overflow][https://api.flutter.dev/flutter/widgets/stack/overflow.html] is not a good idea, and the Flutter team has noticed that it would take time to remove all the usage without breaking Google.
  Since the overflow has been removed in 1.20 and be rolled back now(perhaps 1.22? not for sure). It's hard to say when it will be removed again, so just replace all overflow with clipBehavior.

## [2.5.0]

- Support hook api for tab event. [How to block tab event?](https://github.com/hacktons/convex_bottom_bar/blob/master/doc/how-to-block-tab-event.md)

## [2.4.2]

- Bug fix [#102 dispose is executed more than once](https://github.com/hacktons/convex_bottom_bar/issues/102)
- Set min version of flutter to 1.12.0

## [2.4.1]

- Fix badge alignment issues [#77](https://github.com/hacktons/convex_bottom_bar/issues/77).

## [2.4.0]

- Support flutter v1.20, the v2.2.4-flutter-1.20 is now sync with v2.4.0.
- Add new feature. We can now add corner on AppBars's background.

## [2.3.0]

- Add convex button widget.

## [2.2.5+1]

- fix format issue.

```
lib/src/bar.dart is not formatted according to dartfmt
To format your files run: dartfmt -w .
```

## [2.2.5]

- Bug fix [#67](https://github.com/hacktons/convex_bottom_bar/issues/67)

## [2.2.4-flutter-1.20]

- support the flutter v1.20
- fix: overflow property of stack deprecated [#60](https://github.com/hacktons/convex_bottom_bar/pull/60)

## [2.2.4]

- Polish internal style, avoid reanimate when AppBar rebuild.

## [2.2.3]

- Bug fix [#59](https://github.com/hacktons/convex_bottom_bar/issues/59)
- Bug fix: default index not match with DefaultTabController.

## [2.2.2]

- Fix activate index when appbar is not working with controller;
- Dispose AnimationController when is dirty;

## [2.2.1]

- Support RTL;

## [2.2.0]

- Add new API to hook internal styles. Checkout `StyleProvider` for details;
- Provide configuration to hide Text widget when label is empty;

## [2.1.1]

- Fix ChipBuilder missing issue;
- Polish ci test;

## [2.1.0+1]

- Document all the public APIs;

## [2.1.0]

- Add controller for Appbar to change tab index programmaticlly;
- Support with framework's DefaultTabController and TabController;
- Enable access to the ConvexAppBarState;
- Fix active tab position when using even menus;

## [2.0.3]

- Fix hitTest for active tab;

## [2.0.2]

- Add new config parameter for initial active index;

## [2.0.1]

- Update usage instructions in README.md；
- Add new test cases to improve the code coverage;
- Bug fix;

## [2.0.0]

- Support badge on tab item;
- Constructor update, rename the builder;
- Bug fix;

## [1.4.1]

- Add titled style;
- Remove some redundant widget layers;

## [1.4.0]

- Add flip style;
- Add textIn style;

## [1.3.1+1]

- Bug fix: activate icon is not working in some style;
- Improvement: enable image with/without color blend;

## [1.3.1]

- Support gradient background;
- Tab item are generic type, both `IconData` and `Widget` can be used;

## [1.3.0]

- Add new tab style;
- Support animated tab transition；

## [1.2.0]

- Support iPhoneX' safe area at bottom edge. [#issues/7](https://github.com/hacktons/convex_bottom_bar/issues/7)

## [1.1.0]

- Add elevation attribute.

## [1.0.2+1]

- Documentation fixes.

## [1.0.2]

- Fix maintenance suggestions to get higher score

## [0.0.1]

- Publish the packages to pub.dev


lib/src/bar.dart
/*
 *  Copyright 2020 Chaobin Wu <chaobinwu89@gmail.com>
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  
 *      http://www.apache.org/licenses/LICENSE-2.0
 *  
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'chip_builder.dart';
import 'interface.dart';
import 'item.dart';
import 'painter.dart';
import 'stack.dart' as extend;
import 'style/fixed_circle_tab_style.dart';
import 'style/fixed_tab_style.dart';
import 'style/react_circle_tab_style.dart';
import 'style/react_tab_style.dart';
import 'style/styles.dart';

/// Default size of the curve line.
const double CONVEX_SIZE = 80;

/// Default height of the AppBar.
const double BAR_HEIGHT = 50;

/// Default distance that the child's top edge is inset from the top of the stack.
const double CURVE_TOP = -25;

/// Default size for active tab.
const double ACTION_LAYOUT_SIZE = 60;

/// Default size for active icon in tab.
const double ACTION_INNER_BUTTON_SIZE = 40;

/// Default elevation of [ConvexAppBar].
const double ELEVATION = 2;

/// Tab styles.
enum TabStyle {
  /// Convex shape fixed center, see [FixedTabStyle].
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-fixed.gif)
  fixed,

  /// Convex shape is fixed center with circle, see [FixedCircleTabStyle].
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-fixed-circle.gif)
  fixedCircle,

  /// Convex shape is moved after selection, see [ReactTabStyle].
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-react.gif)
  react,

  /// Convex shape is moved with circle after selection, see [ReactCircleTabStyle].
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-react-circle.gif)
  reactCircle,

  /// Tab icon, text animated with pop transition.
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-textIn.gif)
  textIn,

  /// Similar to [TabStyle.textIn], text first.
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-titled.gif)
  titled,

  /// Tab item is flipped when selected, does not support [flutter web].
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-flip.gif)
  flip,

  /// User defined style
  custom,
}

/// Online example can be found at http://hacktons.cn/convex_bottom_bar_gothomesafe.
///
/// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-theming.png)
class ConvexAppBar extends StatefulWidget {
  /// Tab item builder.
  final DelegateBuilder itemBuilder;

  /// Badge chip builder.
  final ChipBuilder? chipBuilder;

  /// Tab Click handler.
  final GestureTapIndexCallback? onTap;

  /// Tab event notifier, can be used to block tap event.
  final TapNotifier? onTapNotify;

  /// Tab controller to work with [TabBarView] or [PageView].
  final TabController? controller;

  /// Color of the AppBar.
  final Color? backgroundColor;

  /// Draw the background with topLeft and topRight corner; Only work work with fixed style
  ///
  /// ![corner image](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-corner.png)
  final double? cornerRadius;

  /// If provided, backgroundColor for tab app will be ignored.
  ///
  /// ![](https://github.com/hacktons/convex_bottom_bar/raw/master/doc/appbar-gradient.gif)
  final Gradient? gradient;

  /// The initial active index, you can config initialIndex of [TabController] if work with [TabBarView] or [PageView].
  final int? initialActiveIndex;

  /// Disable access of DefaultTabController to avoid unexpected conflict.
  final bool disableDefaultTabController;

  /// Tab count.
  final int count;

  /// Height of the AppBar.
  final double? height;

  /// Size of the curve line.
  final double? curveSize;

  /// The distance that the [actionButton] top edge is inset from the top of the AppBar.
  final double? top;

  /// Elevation for the bar top edge.
  final double? elevation;

  /// The curve to use in the forward direction. Only works when tab style is not fixed.
  final Curve curve;

  /// Construct a new appbar with internal style.
  ///
  /// ```dart
  /// ConvexAppBar(
  ///   items: [
  ///     TabItem(title: 'Tab A', icon: Icons.add),
  ///     TabItem(title: 'Tab B', icon: Icons.near_me),
  ///     TabItem(title: 'Tab C', icon: Icons.web),
  ///   ],
  /// )
  /// ```
  ///
  /// You can also define a custom chipBuilder class.
  /// ```dart
  /// class _ChipBuilder extends ChipBuilder {
  ///  @override
  ///  Widget build(BuildContext context, Widget child, int index, bool active) {
  ///    return Stack(
  ///      alignment: Alignment.center,
  ///      children: <Widget>[
  ///        child,
  ///        Positioned.fill(
  ///          child: Align(
  ///            alignment: Alignment.topRight,
  ///            child: Container(
  ///              margin: EdgeInsets.only(top: 10, right: 10),
  ///              padding: EdgeInsets.only(left: 4, right: 4),
  ///              child: Icon(Icons.access_alarm, color: Colors.redAccent),
  ///            ),
  ///          ),
  ///        )
  ///      ],
  ///    );
  ///    ;
  ///  }
  /// }
  ///```
  /// See also:
  ///
  ///  * [ConvexAppBar.builder], define a custom tab style by implement a [DelegateBuilder].
  ///  * [ConvexAppBar.badge], construct a new appbar with styled badge.
  ConvexAppBar({
    Key? key,
    required List<TabItem> items,
    int? initialActiveIndex,
    bool? disableDefaultTabController,
    GestureTapIndexCallback? onTap,
    TapNotifier? onTabNotify,
    TabController? controller,
    Color? color,
    Color? activeColor,
    Color? backgroundColor,
    Gradient? gradient,
    double? height,
    double? curveSize,
    double? top,
    double? elevation,
    double? cornerRadius,
    TabStyle? style,
    Curve? curve,
    ChipBuilder? chipBuilder,
  }) : this.builder(
          key: key,
          itemBuilder: supportedStyle(
            style ?? TabStyle.reactCircle,
            items: items,
            color: color ?? Colors.white60,
            activeColor: activeColor ?? Colors.white,
            backgroundColor: backgroundColor ?? Colors.blue,
            curve: curve ?? Curves.easeInOut,
          ),
          onTap: onTap,
          onTapNotify: onTabNotify,
          controller: controller,
          backgroundColor: backgroundColor,
          count: items.length,
          initialActiveIndex: initialActiveIndex,
          disableDefaultTabController: disableDefaultTabController ?? false,
          gradient: gradient,
          height: height,
          curveSize: curveSize,
          top: top,
          elevation: elevation,
          cornerRadius: cornerRadius,
          curve: curve ?? Curves.easeInOut,
          chipBuilder: chipBuilder,
        );

  /// Define a custom tab style by implement a [DelegateBuilder].
  ///
  /// ```dart
  /// ConvexAppBar(
  ///   count: 5,
  ///   itemBuilder: Builder(),
  /// )
  ///
  /// class Builder extends DelegateBuilder {
  ///   @override
  ///   Widget build(BuildContext context, int index, bool active) {
  ///     return Text('TAB $index');
  ///   }
  /// }
  /// ```
  const ConvexAppBar.builder({
    Key? key,
    required this.itemBuilder,
    required this.count,
    this.initialActiveIndex,
    this.disableDefaultTabController = false,
    this.onTap,
    this.onTapNotify,
    this.controller,
    this.backgroundColor,
    this.gradient,
    this.height,
    this.curveSize,
    this.top,
    this.elevation,
    this.cornerRadius,
    this.curve = Curves.easeInOut,
    this.chipBuilder,
  })  : assert(top == null || top <= 0, 'top should be negative'),
        assert(initialActiveIndex == null || initialActiveIndex < count, 'initial index should < $count'),
        assert(cornerRadius == null || cornerRadius >= 0, 'cornerRadius must >= 0'),
        super(key: key);

  /// Construct a new appbar with badge.
  ///
  /// {@animation 1010 598 https://github.com/hacktons/convex_bottom_bar/raw/master/doc/badge-demo.mp4}
  ///
  /// [badge] is map with tab items, the value of entry can be either [String],
  /// [IconData], [Color] or [Widget].
  ///
  /// ```dart
  /// ConvexAppBar.badge(
  ///   {3: '99+'},
  ///   items: [
  ///     TabItem(title: 'Tab A', icon: Icons.add),
  ///     TabItem(title: 'Tab B', icon: Icons.near_me),
  ///     TabItem(title: 'Tab C', icon: Icons.web),
  ///   ],
  /// )
  /// ```
  factory ConvexAppBar.badge(
    Map<int, dynamic> badge, {
    Key? key,
    // config for badge
    Color? badgeTextColor,
    Color? badgeColor,
    EdgeInsets? badgePadding,
    EdgeInsets? badgeMargin,
    double? badgeBorderRadius,
    // parameter for appbar
    required List<TabItem> items,
    int? initialActiveIndex,
    bool? disableDefaultTabController,
    GestureTapIndexCallback? onTap,
    TapNotifier? onTabNotify,
    TabController? controller,
    Color? color,
    Color? activeColor,
    Color? backgroundColor,
    Gradient? gradient,
    double? height,
    double? curveSize,
    double? top,
    double? elevation,
    double? cornerRadius,
    TabStyle? style,
    Curve? curve,
  }) {
    DefaultChipBuilder? chipBuilder;
    if (badge.isNotEmpty) {
      chipBuilder = DefaultChipBuilder(
        badge,
        textColor: badgeTextColor ?? Colors.white,
        badgeColor: badgeColor ?? Colors.redAccent,
        padding: badgePadding ?? EdgeInsets.only(left: 4, right: 4),
        margin: badgeMargin ?? EdgeInsets.only(top: 10, right: 10),
        borderRadius: badgeBorderRadius ?? 20,
      );
    }
    return ConvexAppBar(
      key: key,
      items: items,
      initialActiveIndex: initialActiveIndex,
      disableDefaultTabController: disableDefaultTabController ?? false,
      onTap: onTap,
      onTabNotify: onTabNotify,
      controller: controller,
      color: color,
      activeColor: activeColor,
      backgroundColor: backgroundColor,
      gradient: gradient,
      height: height,
      curveSize: curveSize,
      top: top,
      elevation: elevation,
      cornerRadius: cornerRadius,
      style: style,
      curve: curve,
      chipBuilder: chipBuilder,
    );
  }

  @override
  ConvexAppBarState createState() {
    return ConvexAppBarState();
  }
}

/// State of [ConvexAppBar].
class ConvexAppBarState extends State<ConvexAppBar> with TickerProviderStateMixin {
  int? _currentIndex;
  int _warpUnderwayCount = 0;
  Animation<double>? _animation;
  AnimationController? _animationController;
  TabController? _controller;

  int _previousTimestamp = 0;
  static const _TRANSITION_DURATION = 150;

  @override
  void initState() {
    if (widget.cornerRadius != null && widget.cornerRadius! > 0 && !isFixed()) {
      throw FlutterError.fromParts(<DiagnosticsNode>[
        ErrorSummary('ConvexAppBar is configured with cornerRadius'),
        ErrorDescription('Currently the corner only work for fixed style, if you are using '
            'other styles, the convex shape can be broken on the first and last tab item '),
        ErrorHint('You should use TabStyle.fixed or TabStyle.fixedCircle to make the'
            ' background display with topLeft/topRight corner'),
      ]);
    }
    _resetState();
    super.initState();
  }

  void _handleTabControllerAnimationTick() {
    if (_warpUnderwayCount > 0 || _controller == null || !_controller!.indexIsChanging) {
      return;
    }
    if (_controller!.index != _currentIndex) {
      _warpToCurrentIndex();
    }
  }

  Future<void> _warpToCurrentIndex() async {
    if (!mounted) return Future<void>.value();
    final c = _controller;
    if (c == null) {
      return;
    }
    // Workaround for TabController, see https://github.com/hacktons/convex_bottom_bar/issues/59
    var _diff = (c.index - _currentIndex!).abs();
    if (_diff == 1) {
      if (_blockEvent(c.index)) return;
      final previousIndex = c.previousIndex;
      final index = c.index;
      _warpUnderwayCount += 1;
      await animateTo(index, from: previousIndex);
      _warpUnderwayCount -= 1;
      return Future<void>.value();
    }
  }

  /// change active tab index; can be used with [PageView].
  Future<void> animateTo(int index, {int? from}) async {
    var gap = DateTime.now().millisecondsSinceEpoch - _previousTimestamp;
    _updateAnimation(
      from: from ?? _currentIndex,
      to: index,
      duration: Duration(milliseconds: gap < _TRANSITION_DURATION ? 0 : _TRANSITION_DURATION),
    );
    // ignore: unawaited_futures
    _animationController?.forward();
    if (mounted) {
      setState(() {
        _currentIndex = index;
      });
    }
    _previousTimestamp = DateTime.now().millisecondsSinceEpoch;
  }

  Animation<double> _updateAnimation({int? from, int? to, Duration duration = const Duration(milliseconds: _TRANSITION_DURATION)}) {
    if (from != null && (from == to) && _animation != null) {
      return _animation!;
    }
    from ??= _controller?.index ?? widget.initialActiveIndex ?? 0;
    to ??= from;
    final lower = (2 * from + 1) / (2 * widget.count);
    final upper = (2 * to + 1) / (2 * widget.count);
    _animationController?.dispose();
    final controller = AnimationController(duration: duration, vsync: this);
    final curve = CurvedAnimation(
      parent: controller,
      curve: widget.curve,
    );
    _animationController = controller;
    return _animation = Tween(begin: lower, end: upper).animate(curve);
  }

  @override
  void dispose() {
    _controller?.animation?.removeListener(_handleTabControllerAnimationTick);
    _controller = null;

    _animationController?.dispose();
    super.dispose();
  }

  TabController? get _takeControllerRef {
    if (widget.disableDefaultTabController == true) {
      return widget.controller;
    }
    return widget.controller ?? DefaultTabController.of(context);
  }

  void _updateTabController() {
    final newController = _takeControllerRef;
    assert(() {
      if (newController != null && widget.controller == null && widget.initialActiveIndex != null) {
        throw FlutterError('ConvexAppBar.initialActiveIndex is not allowed when working with TabController.\n'
            'Please setup through TabController.initialIndex, or disable DefaultTabController by #disableDefaultTabController');
      }
      return true;
    }());
    if (newController == _controller) return;
    _controller?.animation?.removeListener(_handleTabControllerAnimationTick);
    _controller = newController;
    _controller?.animation?.addListener(_handleTabControllerAnimationTick);
  }

  void _resetState() {
    var index = _controller?.index ?? widget.initialActiveIndex;
    // when both initialActiveIndex and controller are not configured
    _currentIndex = index ?? 0;

    if (!isFixed() && _controller != null) {
      // when controller is not defined, the default index can rollback to 0
      // https://github.com/hacktons/convex_bottom_bar/issues/67
      _updateAnimation();
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (_controller != _takeControllerRef) {
      _updateTabController();
      _resetState();
    }
  }

  @override
  void didUpdateWidget(ConvexAppBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.controller != oldWidget.controller || widget.count != oldWidget.count) {
      _updateTabController();
      _resetState();
    }
  }

  @override
  Widget build(BuildContext context) {
    // take care of iPhoneX' safe area at bottom edge
    final additionalBottomPadding = math.max(MediaQuery.of(context).padding.bottom, 0.0);
    final convexIndex = isFixed() ? (widget.count ~/ 2) : _currentIndex;
    final active = isFixed() ? convexIndex == _currentIndex : true;

    final height = (widget.height ?? BAR_HEIGHT) + additionalBottomPadding;
    final width = MediaQuery.of(context).size.width;
    var percent = isFixed() ? const AlwaysStoppedAnimation<double>(0.5) : _animation ?? _updateAnimation();
    var factor = 1 / widget.count;
    var textDirection = Directionality.of(context);
    var dx = convexIndex! / (widget.count - 1);
    if (textDirection == TextDirection.rtl) {
      dx = 1 - dx;
    }

    var offset = FractionalOffset(widget.count > 1 ? dx : 0.0, 0);
    return extend.Stack(
      clipBehavior: Clip.none,
      alignment: Alignment.bottomCenter,
      children: <Widget>[
        Container(
          height: height,
          width: width,
          child: CustomPaint(
            painter: ConvexPainter(
              top: widget.top ?? CURVE_TOP,
              width: widget.curveSize ?? CONVEX_SIZE,
              height: widget.curveSize ?? CONVEX_SIZE,
              color: widget.backgroundColor ?? Colors.blue,
              gradient: widget.gradient,
              sigma: widget.elevation ?? ELEVATION,
              leftPercent: percent,
              textDirection: textDirection,
              cornerRadius: widget.cornerRadius,
            ),
          ),
        ),
        _barContent(height, additionalBottomPadding, convexIndex),
        Positioned.fill(
          top: widget.top ?? CURVE_TOP,
          bottom: additionalBottomPadding,
          child: FractionallySizedBox(
              widthFactor: factor,
              alignment: offset,
              child: GestureDetector(
                onTap: () => _onTabClick(convexIndex),
                child: _newTab(convexIndex, active),
              )),
        ),
      ],
    );
  }

  /// Whether the tab shape are fixed or not.
  bool isFixed() => widget.itemBuilder.fixed();

  Widget _barContent(double height, double paddingBottom, int curveTabIndex) {
    var children = <Widget>[];
    for (var i = 0; i < widget.count; i++) {
      if (i == curveTabIndex) {
        children.add(Expanded(child: Container()));
        continue;
      }
      var active = _currentIndex == i;
      children.add(Expanded(
        child: GestureDetector(
          behavior: HitTestBehavior.opaque,
          onTap: () => _onTabClick(i),
          child: _newTab(i, active),
        ),
      ));
    }

    return Container(
      height: height,
      padding: EdgeInsets.only(bottom: paddingBottom),
      child: Row(
        mainAxisSize: MainAxisSize.max,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: children,
      ),
    );
  }

  Widget _newTab(int i, bool active) {
    final child = widget.itemBuilder.build(context, i, active);
    return widget.chipBuilder?.build(context, child, i, active) ?? child;
  }

  bool _blockEvent(int i) {
    if (widget.onTapNotify != null && !widget.onTapNotify!(i)) {
      debugPrint('tap event block by ${widget.onTapNotify}');
      return true;
    }
    return false;
  }

  void _onTabClick(int i) {
    if (_blockEvent(i)) return;
    animateTo(i);
    _controller?.animateTo(i);
    widget.onTap?.call(i);
  }

  /// Used to simulate tab event on tab item; This will notify [ConvexAppBar.onTap];
  ///
  /// Also see:
  /// * [animateTo]
  void tap(int index) {
    _onTabClick(index);
  }
}

/// Hook for internal tab style. Unlike the [ConvexAppBar.builder], you may want to
/// update the tab style without define a new tab style.
///
/// Warning:
/// This hook is limited, and can lead to `overflow broken` if the size you provide
/// does no match with internal style.
class StyleProvider extends InheritedWidget {
  /// Style configuration
  final StyleHook style;

  /// Provide style to provider, [ConvexAppBar] will bind to the provided style.
  /// See also:
  ///
  ///  * [ConvexAppBar]
  ///  * [StyleHook]
  StyleProvider({Key? key, required this.style, required Widget child}) : super(key: key, child: child);

  /// Get instance of style provider, can be null if you're not providing one.
  static StyleProvider? of(BuildContext context) {
    return context.dependOnInheritedWidgetOfExactType<StyleProvider>();
  }

  @override
  bool updateShouldNotify(StyleProvider oldWidget) {
    return style.activeIconMargin != oldWidget.style.activeIconMargin || style.activeIconSize != oldWidget.style.activeIconSize || style.iconSize != oldWidget.style.iconSize;
  }
}
